Update collector to v0.104.0

mackjmr opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?
This PR updates all collector dependencies to v0.104.0 except datadogconnector. datadogconnector cannot be upgraded because of breaking change in #26347.
<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes
Spin up agent with OTLP ingest with defaults and make sure the http/ grpc servers addresses are `localhost` instead of `0.0.0.0`.

mackjmr committed:
Update collector to v0.104.0

This PR updates all collector dependencies to v0.104.0 except datadogconnector. datadogconnector cannot be upgraded because of breaking change in #26347.

songy23 commented on go.mod:
```diff
@@ -1349,6 +1350,9 @@ exclude (
 	github.com/knadh/koanf/providers/confmap v0.1.0
 )
 
+// todo: update datadog connector with breaking changes from https://github.com/DataDog/datadog-agent/pull/26347.
```
cc @ajgajg1134 with the `replace` this PR shouldn't block  https://github.com/DataDog/datadog-agent/pull/26347, but please let us know if you ran into dependency issues.

songy23 reviewed (commented):
As suggested by @mx-psi please add a release note

mackjmr committed:
add release note

mackjmr committed:
fix ci

dineshg13 reviewed (changes requested):
Please don't edit comp/otelcol/collector-contrib/impl/go.mod directly. Instead update in 
`comp/otelcol/collector-contrib/impl/manifest.yaml` and run `inv collector.generate` . 

Also please update the version here https://github.com/DataDog/datadog-agent/blob/d8bdcaf6208f1d3d36d5b8a4ae1dc9d65924b64e/comp/otelcol/collector/impl/collector.go#L94

dineshg13 commented:
We do need a QA to ensure the upstream breaking changes aren't impacting OTLP ingest in the Agent

mackjmr committed:
inv generate-licenses

mackjmr committed:
inv collector.generate

mackjmr commented:
@dineshg13 I updated `manifest.yaml` and version in `collector.yaml`. I also added QA instructions, although I don't think there is anything to check other than the http/ grpc servers addresses are `localhost` instead of `0.0.0.0`.

dineshg13 reviewed (approved):
We are using datadog connector `0.104.0` , but otherwise LGTM

songy23 commented:
> We are using datadog connector 0.104.0 , but otherwise LGTM

It is replaced with 0.103.0 so shouldn't be a problem

GustavoCaso commented on releasenotes/notes/collector-104-8697c1d5cc4b1745.yaml:
```diff
@@ -0,0 +1,11 @@
+---
+upgrade:
+  - |
+    Change default bind address in OTLP ingest from `0.0.0.0` to `localhost`. Please refer to the following blog post: https://opentelemetry.io/blog/2024/hardening-the-collector-one for additional information around this change.
```
The release note has nothing to do with the title of the PR, is that intended?

mackjmr replied:
The bump to v0.104.0 brings in the change mentioned in the changelog, see: https://github.com/open-telemetry/opentelemetry-collector/releases/tag/v0.104.0.

GustavoCaso replied:
Nit: Maybe mention the upgrade to `0.104.0` of the collector

GustavoCaso commented on comp/otelcol/collector-contrib/def/go.mod:
```diff
@@ -34,32 +34,26 @@ require (
 	github.com/shirou/gopsutil/v4 v4.24.5 // indirect
 	github.com/shoenig/go-m1cpu v0.1.6 // indirect
-	github.com/spf13/cobra v1.8.0 // indirect
+	github.com/spf13/cobra v1.8.1 // indirect
```
Is this bump intended?

mackjmr committed:
Merge branch 'main' into mackjmr/update-collector-104

mackjmr committed:
downgrade datadogconnector to v0.103.0 in manifest

mackjmr commented:
> @dineshg13 @songy23
We are using datadog connector 0.104.0 , but otherwise LGTM
Are you talking about the version in manifest.yaml ? I can bump this down to 0.103.0 ?

I didn't realize `inv collector.generate` was removing the replace statement in `comp/otelcol/collector-contrib/impl/go.mod`. I've updated the datadogconnector in manifest.yaml to v0.103.0 with a todo. I've also updated the OCB version to v0.104.0 in manifest.yaml.

mackjmr committed:
Merge branch 'main' into mackjmr/update-collector-104

amenasria reviewed (approved):
LGTM for Agent DevX Infra file !

mackjmr committed:
use replace in manifest

mackjmr committed:
Merge main

mackjmr committed:
Revert "Merge main"

This reverts commit 3030da4838b86cdea6ada8b5a21a06f5d26b7f02.

mackjmr committed:
Merge branch 'main' into mackjmr/update-collector-104

mackjmr committed:
fix

mackjmr committed:
fix ci

mackjmr committed:
fix lint

mackjmr committed:
fix test

hghotra reviewed (approved):
👍🏼 - Serverless

mackjmr commented:
/merge

KSerrania commented:
/merge
